torch/csrc/distributed/c10d/init.cpp

```diff
@@ -487,18 +487,25 @@ An enum-like class for built-in communication hooks: ``ALLREDUCE`` and ``FP16_CO
       py::call_guard<py::gil_scoped_release>());
 
   py::enum_<::c10d::ReduceOp>(module, "ReduceOp", R"(
-An enum-like class for available reduction operations: ``SUM``, ``PRODUCT``,
-``MIN``, ``MAX``, ``BAND``, ``BOR``, and ``BXOR``.
+An enum-like class for available reduction operations: ``SUM``, ``AVG``,
+``PRODUCT``, ``MIN``, ``MAX``, ``BAND``, ``BOR``, and ``BXOR``.
 
-Note that ``BAND``, ``BOR``, and ``BXOR`` reductions are not available when
+``BAND``, ``BOR``, and ``BXOR`` reductions are not available when
 using the ``NCCL`` backend.
 
+``AVG`` divides values by the world size before summing across ranks.
+``AVG`` is only available with the ``NCCL`` backend,
+and only for NCCL versions 2.10 or later.
+
 Additionally, ``MAX``, ``MIN`` and ``PRODUCT`` are not supported for complex tensors.
 
 The values of this class can be accessed as attributes, e.g., ``ReduceOp.SUM``.
 They are used in specifying strategies for reduction collectives, e.g.,
 :func:`reduce`, :func:`all_reduce_multigpu`, etc.)")
       .value("SUM", ::c10d::ReduceOp::SUM)
+#ifdef NCCL_HAS_AVG
+      .value("AVG", ::c10d::ReduceOp::AVG)
+#endif
```

Comment:
Please change it to a dymanic version check, something like:
```
auto& reduce_op_enum = py::enum_<::c10d::ReduceOp>(module, "ReduceOp", " ....").value().value(...);
if (torch::cuda::nccl::version() >= 3000 ) // NCCL version is greater than 3.10
   reduce_op_enum.value("AVG", ::c10d::ReduceOp::AVG);
```

Reply:
Hmm, the runtime check `torch::cuda::nccl::version()` is declared in `torch/csrc/cuda/nccl.h`.

init.cpp defines the Python bindings. Based on https://github.com/pytorch/pytorch/pull/62835#discussion_r683775243 it sounds like we'd rather not include `torch/csrc/cuda/nccl.h` here in init.cpp, which means we'd rather not use `torch::cuda::nccl::version()` Am i understanding correctly?

I submitted an alternate approach that exposes
```c++
.value("AVG", ::c10d::ReduceOp::AVG)
```
unconditionally, but throws a helpful error in `ProcessGroupNCCL.cpp` if ncclAvg isn't actually available.

Reply:
I used macros. Maybe we _still_ want a runtime check in case the pytorch was built with some earlier nccl and dynamically links against a newer version, but that seems potentially dangerous in general and maybe not a use case we want to support.

Reply:
Never mind, based on slack conversations, it actually _is_ ok to include `torch/csrc/cuda/nccl.h` and use `torch::cuda::nccl::version()` in init.cpp.

Afaict my "alternative approach" is also alright, though.

Reply:
wait a minute, `torch::cuda::nccl::version()` isn't really a runtime check on the (potentially dynamically linked) nccl version at all. It's just a wrapper that packages the NCCL_MAJOR/MINOR/PATCH macros. So using the macros directly in ProcessGroupNCCL.cpp is no worse or better.